### PR TITLE
LMDK: Loadable Modules Dev Kit build scripts

### PR DESCRIPTION
**Pull request by serhiy-katsyuba-intel**
Initial commit with set of cmake and linker scripts to build loadable modules binary.

Things yet TODO:
  add all necessary Dev Kit header files to include dir,
  replace dummy example with somewhat useful module example.

### CODE REVIEW

**Comment by serhiy-katsyuba-intel** [comment]
I will be on leave till April 10, so no response until then.

File: lmdk/cmake/ldscripts/bss_linker_script.txt
Comment (lgirdwood) [inline lmdk/cmake/ldscripts/bss_linker_script.txt:1]: Can we have a comment in each linker script about what it does and how it's used. i.e. module developers should be able to figure out the scripts used for thier use case and integrate/debug features as needed.

File: lmdk/cmake/xtensa.cmake
Comment (lgirdwood) [inline lmdk/cmake/xtensa.cmake:13]: GCC should also work here.

File: lmdk/modules/dummy/dummy.c
Comment (lgirdwood) [inline lmdk/modules/dummy/dummy.c:8]: Lets use the correct code style.

File: lmdk/modules/dummy/dummy.c
Comment (lgirdwood) [inline lmdk/modules/dummy/dummy.c:22]: and correct code style here too.

File: lmdk/readme.txt
Comment (lgirdwood) [inline lmdk/readme.txt:2]: Needs to be called README.md for GH to render.

**Review by lgirdwood — COMMENTED** [review]
Initial review, needs to documentation to move forward.

**Review by aborisovich — CHANGES_REQUESTED** [review]
There are some things that we need to address, but overall it looks quite good.
My biggest concern is a editing manually config.cmake file. I assume this was just done for testing to show us how the draft works. See comments below

**Comment by sys-pt1s** [comment]
Can one of the admins verify this patch?

File: lmdk/cmake/ldscripts/bss_linker_script.txt
Comment (serhiy-katsyuba-intel) [reply lmdk/cmake/ldscripts/bss_linker_script.txt:1]: Added description about linker scripts to ldscripts.cmake.

File: lmdk/modules/dummy/dummy.c
Comment (serhiy-katsyuba-intel) [reply lmdk/modules/dummy/dummy.c]: Fixed.

File: lmdk/readme.txt
Comment (serhiy-katsyuba-intel) [reply lmdk/readme.txt]: Fixed.

File: lmdk/cmake/build.cmake
Comment (serhiy-katsyuba-intel) [reply lmdk/cmake/build.cmake:2]: Done.

File: lmdk/cmake/build.cmake
Comment (serhiy-katsyuba-intel) [reply lmdk/cmake/build.cmake]: Fixed. Now cmake -E cat is used.
Actually, it worked on Windows. Maybe it was using cat from MinGW, not sure.

File: lmdk/readme.txt
Comment (serhiy-katsyuba-intel) [reply lmdk/readme.txt]: Fixed. Now SIGNING_KEY and RIMAGE_COMMAND are expected to be specified at cmake command line and were removed from config.cmake.

File: lmdk/cmake/build.cmake
Comment (serhiy-katsyuba-intel) [reply lmdk/cmake/build.cmake]: Removed that --verbose at all as seems it is not really useful.

File: lmdk/cmake/config.cmake
Comment (serhiy-katsyuba-intel) [reply lmdk/cmake/config.cmake]: Fixed. Removed it from config.cmake. Now SIGNING_KEY is supplied as cmake command line parameter.

File: lmdk/libraries/dummy/CMakeLists.txt
Comment (serhiy-katsyuba-intel) [reply lmdk/libraries/dummy/CMakeLists.txt]: Updated CMake required version to 3.20.

File: lmdk/cmake/xtensa.cmake
Comment (serhiy-katsyuba-intel) [reply lmdk/cmake/xtensa.cmake]: Renamed toolchain file to xtensa-toolchain.cmake. The file is not included but used as
```
set(CMAKE_TOOLCHAIN_FILE "${CMAKE_CURRENT_LIST_DIR}/../../cmake/xtensa-toolchain.cmake")
```
in dummy library CMakeLists.txt. So if that line is commented out, the toolchain file may be used with cmake --tollchain=<path> option.

**Comment by pjdobrowolski** [comment]
Building modules required fresh rimage tool, is it a good idea to use west tool for checking freshness of repositories and build?

**Comment by kv2019i** [comment]
@pjdobrowolski wrote:
> Building modules required fresh rimage tool, is it a good idea to use west tool for checking freshness of repositories and build?

SOF rimage is fixed to version in sof/west.yml. So if you need a newer rimage, the rimage version needs to be upgraded first in SOF main, and then you can merge your PR and trust the rimage is always new enough.

**Comment by lgirdwood** [comment]
@serhiy-katsyuba-intel can we track the rimage & west dependencies here so we can merge after deps are merged.

File: lmdk/README.md
Comment (kv2019i) [inline lmdk/README.md:3]: @serhiy-katsyuba-intel  I think this TODO and the TODO in commit is confusing a bit. Is this PR intended for merging now, or do you plan to work on this still before merging?

Comment (pjdobrowolski) [reply lmdk/README.md:3]: I've made an introduction part in sof-docs https://github.com/thesofproject/sof-docs/pull/463, it is lacking architecture part, but I'm going to fill it up after May break.

**Comment by serhiy-katsyuba-intel** [comment]
@aborisovich , I've addressed the changes you requested, could you redo the review ("changes requested" status blocks merging)?

**Comment by aborisovich** [comment]
> @aborisovich , I've addressed the changes you requested, could you redo the review ("changes requested" status blocks merging)?

I'll check those in a moment and let you know

File: lmdk/cmake/xtensa-toolchain.cmake
Comment (aborisovich) [inline lmdk/cmake/xtensa-toolchain.cmake:6]: Started testing this PR and found few issues.
Issue 1: when XTENSA_TOOLCHAIN_PATH variable is not set, we got following error:

> CMake Error at C:/Workspace/my_sof_workspace/sof/lmdk/cmake/xtensa-toolchain.cmake:15 (cmake_path):
>   cmake_path CONVERT must be called with three or four arguments.

This is not very elegant, proper error message should be displayed.
Issue 2: toolchain is ignored and other installed is picked if anything fails in xtensa-toolchain.cmake script.
repro steps:
1. `cd sof\lmdk\libraries\dummy`
2. `set XTENSA_TOOLCHAIN_PATH=qweqew`
3. `cmake -S . -B build -G Ninja --toolchain "..\..\cmake\xtensa-toolchain.cmake" -DRIMAGE_COMMAND="qweqew" -DSIGNING_KEY="whateverkey"`
Result:
[cmake.log](https://github.com/thesofproject/sof/files/11390723/cmake.log)

As you can see, if we do not file any error message in the toolchain file, other installed toolchain on the end-user computer is taken (what is certainly not desired).

**Review by aborisovich — CHANGES_REQUESTED** [review]
During testing, some issues found. I know it is dummy but still we should make sure it works in some expected way right?

File: lmdk/cmake/xtensa-toolchain.cmake
Comment (aborisovich) [inline lmdk/cmake/xtensa-toolchain.cmake:25]: What is the purpose of those?

Comment (serhiy-katsyuba-intel) [reply lmdk/cmake/xtensa-toolchain.cmake:25]: These are path to sysroot for cross-compilation and settings how to search for execs, headers and libraries. CMake uses these settings to alter find_library() behaviour. These here are set to search only in sysroot provided by toolchain and never search in host sysroot. For example, if find_library() is called to search for e.g. libm.a, without these settings it can try to find libm.a in your host GCC installation but we need libm.a that comes with cross-platform toolchain.

**Comment by serhiy-katsyuba-intel** [comment]
Changes in the updated version:
* Proper error message is displayed if XTENSA_TOOLCHAIN_PATH environment or CMake variable not defined.
* Now if required toolchain not found, CMake will NOT try to find some other toolchain to use.
* Indentation in .cmake files replaced to use tabs.

**Comment by serhiy-katsyuba-intel** [comment]
Just in case, the error reported by checkpatch
```
ERROR: trailing whitespace
#68: FILE: lmdk/README.md:10:
```
is kind of false positive: there intentionally has to be a spaces in README.md.

3 warnings in checkpatch
```
Prefer __section(".buildinfo") over __attribute__((section(".buildinfo")))
```
not worth to be fixed: currently dummy loadable module does not include any header files. In future, proper headers should be added for loadable modules and these could have such macro defined. Also dummy loadable module example in future will be replaced with more practical example.

**Review by aborisovich — APPROVED** [review]
All my questions had been answered, looks good now.

**Review by aborisovich — CHANGES_REQUESTED** [review]
Sorry, found the reproduction where it still does not work (rimage issue). My previous request changes was not addressed. I've prepared a fix if you like - just cherry pick https://github.com/aborisovich/sof/commit/eafbaa2469b411f99c15e87cbaf5103d9cd36fc5 or if you don't fix it differently. Either way I think passing a relative path to rimage should work.
image

**Review by aborisovich — APPROVED** [review]
We agreed to merge this as is without relative path to rimage working and I will provide this fixup in separate PR.

**Review by pjdobrowolski — APPROVED** [review]

**Review by softwarecki — APPROVED** [review]

**Comment by kv2019i** [comment]
Ready to go, need a clean CI run still.

**Pull request merged**
